src/CoercibleUtils.hs

```diff
@@ -1,8 +1,58 @@
-module CoercibleUtils where
+{- |
+Utility functions for Coercible types.
+Primarily pulled from the package `newtype-generics`, and
+based on Conor McBride's Epigram work, but
+generalised to work over anything `Coercible`.
 
-import Data.Coerce
+>>> ala Sum foldMap [1,2,3,4 :: Int] :: Int
+10
+
+>>> ala Endo foldMap [(+1), (+2), (subtract 1), (*2) :: Int -> Int] (3 :: Int) :: Int
+8
+
+>>> under2 Min (<>) 2 (1 :: Int) :: Int
+1
+
+>>> over All not (All False) :: All
+All {getAll = True)
+
+Users might also find the GHC plugin <https://github.com/mpickering/hashtag-coerce hashtag-coerce> useful in tandem with this library.
+-}
+module CoercibleUtils
+  ( (#.), (.#)
+  , op
+  , ala, ala'
+  , under, over
+  , under2, over2
+  , underF, overF
+  ) where
+
+import Data.Coerce (Coercible, coerce)
 
 -- | Coercive left-composition.
+--
+--   The coercive composition operators solve a problem as presented in <https://ghc.haskell.org/trac/ghc/ticket/7542 GHC Trac #7542>.
+--
+--   The problem, in a nutshell:
+--
+--   If N is a newtype constructor, then N x will always have the same
+--   representation as x (something similar applies for a newtype deconstructor).
+--   However, if f is a function,
+--   
+--   N . f = \x -> N (f x)
+--   
+--   This looks almost the same as f, but the eta expansion lifts it--the lhs could
+--   be `_|_`, but the rhs never is. This can lead to very inefficient code. Thus we
```

Comment:
I'm wondering if "the rhs never is [bottom]" is really correct in this example. Does "rhs" refer to `\x -> N (f s)`?

Reply:
Yes. the RHS is never bottom because it is wrapped by N

Reply:
I'm still failing to properly understand it. I even asked about it on \#ghc:

```
(23:23:54) sjakobi: I have some difficulty understanding the Note [Function coercion] in Data.Functor.Utils, in particular this bit:
(23:23:57) sjakobi: N . f = \x -> N (f x)
(23:23:57) sjakobi: This looks almost the same as f, but the eta expansion lifts it--the lhs could
(23:23:57) sjakobi: be _|_, but the rhs never is.
(23:25:03) sjakobi: The RHS just seems to be the definition of the LHS, so I don't see how one could be bottom but the other not.
(23:30:49) geekosaur: sjakobi, it's not because you're explicitly calling for the RHS to be transformed in such a way that _|_ will explode there; the LHS can be transformed in ways that don't necessarily do so. (we are transforming the program source here, not running the code)
(23:31:05) geekosaur: at least, I think that's the point
(23:33:11) geekosaur: so the translation doesn't alter strictness and does preserve other properties that allow transformation to more efficient forms, instead of forcing further transformations to preserve a potential runtime _|_ introduced by lifting so strictness isn't altered
(23:36:41) sjakobi: geekosaur: Thanks, but I can't follow :/                                                                                                                                                            
(23:37:06) geekosaur: later transformations can't tell that the lambda was introduced by an earlier one and can think it changes the strictness, as I read it
(23:37:21) geekosaur: whereas in the cases where it's applied, it doesn't/can't
(23:39:28) sjakobi: Hm, seems like I should read up on strictness analysis. :)
(23:40:12) geekosaur: it's beyond even that; you want transformations after strictness analysis to preserve the determined strictness. this looks like it doesn't, even though in reality it does
(23:40:13) sjakobi: To me it looked like a bad analogy to the inequality f /= \x -> f x
(23:40:30) geekosaur: if I am understanding it correctly, at least
(23:40:49) geekosaur: which can cause later stages to preserve what they think is the correct strictness
(23:41:41) geekosaur: there have been … lots of bugs in this area, things that alter strictness unintentionally either directly or by misleading other transformations into doing so
(23:42:19) geekosaur: (stranal applies specifically to function parameters, not necessarily within expressions as they are transformed)
(23:46:17) geekosaur: that said, I could well be misreading this; I'm still rather new to ghc internals myself
(23:48:13) sjakobi: Thanks, geekosaur. If your explanation is correct, I can't really understand why the note doesn't mention the transformations you speak of.
(23:49:09) geekosaur: there's probably too many of them. every transformation interacts in some way with every other, which is how bugs like the one mentioned creep in
(23:49:27) geekosaur: also, likely that bug contains more detail, which is why it's linked there
(23:55:17) sjakobi: Right, a comment on the ticket mentions Note [Dealing with bottom] in CoreArity, where I can start to dig deeper! :) Thanks for now!
```

I'm kind of opposed to publishing something that I don't understand myself. And admittedly I want this library to be useful to people who might have an even worse understanding of strictness in GHC than mine.

I'm wondering if we should remove the note for now and maybe just give the semantics with respect to bottom like [`Data.Profunctors.Unsafe`](http://hackage.haskell.org/package/profunctors-5.2.2/docs/Data-Profunctor-Unsafe.html#v:-35-.):

```
(#.) ≡ \packer -> \f -> f `seq` packer . f
(.#) ≡ \f -> f `seq` \packer -> f . packer
```